### DISCUSSION TITLE
트랜잭션 Isolation Level에 대한 설명

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#189] [Database]
.

**Comment by alstn113** [comment]
데이터베이스에서 Isolation Level(격리 수준)은 트랜잭션이 동시에 실행될 때 데이터의 일관성과 격리를 보장하는 방법을 정의합니다.
각 수준은 트랜잭션 간 상호작용을 어느 정도까지 허용할지를 설정하여 데이터 무결성을 유지하고 성능을 조절합니다.
주요 격리 수준에는 다음과 같은 네 가지가 있습니다:

> Dirty Read: 하나의 트랜잭션이 커밋되지 않은 다른 트랜잭션의 변경 내용을 읽는 현상
> Non-repeatable Read: 동일 트랜잭션 내에서 두 번 읽는 데이터가 다를 수 있는 현상(하나의 레코드)
> Phantom Read: 동일 트랜잭션 내에서 범위 쿼리를 두 번 실행할 때, 그 사이에 데이터를 삽입하거나 삭제하여 데이터 집합이 달라지는 현상(레코드 집합)

### 1. Read Uncommitted

- 설명: 트랜잭션이 커밋되지 않은 데이터를 읽을 수 있는 가장 낮은 격리 수준입니다.
- 특징:
    - Dirty Read 발생 가능: 다른 트랜잭션이 커밋되지 않은 데이터를 읽을 수 있어 데이터 불일치 가능.
    - Non-repeatable Read와 Phantom Read도 발생 가능.
- 장점: 가장 빠르지만 데이터 무결성 측면에서 위험이 큼.
- 사용 예시: 데이터 일관성보다 성능이 더 중요한 경우.

### 2. Read Committed

- 설명: 트랜잭션이 커밋된 데이터만 읽을 수 있도록 보장합니다.
- 특징:
    - Dirty Read 방지.
    - Non-repeatable Read 발생 가능: 동일 트랜잭션 내에서 두 번 읽는 데이터가 다를 수 있음.
    - Phantom Read도 발생 가능.
- 장점: 데이터 일관성을 어느 정도 유지하면서도 성능이 적절하게 유지됨.
- 사용 예시: 대부분의 관계형 데이터베이스의 기본 설정.

### 3. Repeatable Read

- 설명: 동일 트랜잭션 내에서 동일한 데이터를 여러 번 읽어도 항상 같은 결과를 보장합니다.
- 특징:
    - Dirty Read와 Non-repeatable Read 방지.
    - Phantom Read 발생 가능: 트랜잭션 중 새로운 데이터가 삽입되거나 삭제될 수 있음.
    - InnoDB의 경우 (Gap Lock과 Next-Key Lock을 사용하여) Phantom Read 방지 가능하나 잠금을 동반한 SELECT에서는 발생 가능.
- 장점: 높은 수준의 데이터 일관성을 제공.
- 사용 예시: 데이터의 일관성이 중요한 애플리케이션.

### 4. Serializable

- 설명: 가장 높은 격리 수준으로, 트랜잭션을 순차적으로 실행하여 트랜잭션 간 상호작용을 완전히 차단합니다.
- 특징:
    - Dirty Read, Non-repeatable Read, Phantom Read 모두 방지.
    - 트랜잭션 간 완벽한 격리 보장.
- 장점: 최고 수준의 데이터 일관성을 제공.
- 단점: 성능 저하 가능성이 가장 큼.
- 사용 예시: 금융 애플리케이션 등 데이터 일관성이 매우 중요한 경우.

### 격리 수준의 요약

- Read Uncommitted: 데이터 불일치 가능성이 높지만 성능이 빠름.
- Read Committed: Dirty Read 방지, 일반적인 격리 수준.
- Repeatable Read: Dirty Read와 Non-repeatable Read 방지, 높은 일관성.
- Serializable: 모든 이상 현상 방지, 최고 수준의 일관성, 성능 저하.


참고: https://velog.io/@hgo641/트랜잭션-격리-수준#:~:text=즉%2C%20NON-REPEATABLE%20READ%20는,때%20발생하는%20문제이다.

**Comment by 3Juhwan** [comment]
트랜잭션의 격리 수준은 동시에 여러 트랜잭션이 실행될 때, 한 트랜잭션이 다른 트랜잭션의 연산에 영향을 받지 않도록 하는 정도를 말합니다. 낮은 격리 수준은 동시 처리 능력을 높이지만, 데이터의 일관성 문제를 발생시킬 수 있습니다. 반면, 높은 격리 수준은 데이터의 일관성을 보장하지만, 동시 처리 능력이 떨어질 수 있습니다. 즉, 데이터 정합성과 성능은 반비례합니다. 트랜잭션 isolation level은 개발자가 트랜잭션 격리 수준을 설정할 수 있는 기능을 제공하는 기능입니다. 

**READ UNCOMMITTED**

- 커밋이 되지 않은 트랜잭션의 데이터 변경 내용을 다른 트랜잭션이 조회하는 것을 허용합니다. 
- Dirty Read, Phantom Read, Non-Repeatable Read 문제를 발생시킬 수 있습니다.

**READ COMMITTED**

- 커밋이 완료된 트랜잭션의 변경사항만 다른 트랜잭션에서 조회할 수 있도록 허용합니다. 
- 특정 트랜잭션이 이루어지는 동안, 다른 트랜잭션은 해당 데이터에 접근할 수 없습니다. 
- Phantom Read, Non-Repeatable Read 문제를 발생시킬 수 있습니다. 

**REPEATABLE READ**

- 한 트랜잭션에서 특정 레코드을 조회시 항상 같은 데이터를 응답하는 것을 보장합니다. 
- 하지만, SERIALIZABLE과 다르게 행이 추가되는 것을 막지는 않습니다. 
- Phantom Read 문제를 발생시킬 수 있습니다. 
- InnoDB의 기본 격리 수준입니다. 

**SERIALIZABLE**

- 특정 트랜잭션이 사용중인 테이블의 모든 행을 다른 트랜잭션이 접근할 수 없도록 잠급니다. 
- 가장 높은 데이터 정합성을 갖으나, 성능은 가장 떨어집니다.  
- 단순한 SELECT 쿼리가 실행되더라도, 데이터베이스 락이 걸려 다른 트랜잭션에서 데이터에 접근할 수 없습니다. 

</br>

다음은 트랜잭션 격리 수준에 따라 발생할 수 있는 문제점입니다. 

**Dirty Read**

한 트랜잭션이 다른 트랜잭션이 변경 중인 데이터를 읽는 경우 발생합니다. 다른 트랜잭션이 아직 커밋되지 않은 (즉, 롤백할 가능성이 있는) 데이터를 읽어서, 그 데이터가 나중에 롤백될 경우 트랜잭션의 결과가 변경될 수 있습니다. 이는 데이터의 일관성을 깨뜨릴 수 있습니다.

**Phantom Read**

한 트랜잭션이 동일한 쿼리를 두 번 실행했을 때, 두 번의 쿼리 사이에 다른 트랜잭션이 삽입, 갱신, 삭제 등의 작업을 수행하여 결과 집합이 달라지는 경우를 말합니다. 이로 인해 한 트랜잭션 내에서 일관성이 없는 결과를 가져올 수 있습니다.

**Non-Repeatable Read**

같은 트랜잭션 안에서 동일한 쿼리를 실행했을 때, 다른 결과를 얻는 경우를 의미합니다. 예를 들어, 한 트랜잭션이 같은 데이터를 두 번 읽을 때, 첫 번째 읽기와 두 번째 읽기 사이에 다른 트랜잭션이 해당 데이터를 변경했을 경우 발생할 수 있습니다.


https://hudi.blog/transaction-isolation-level/

**Comment by le2sky** [comment]
변경하거나 조회하고 있는 데이터에 대해서 다른 트랜잭션이 볼 수 있는지 여부 결정
동시 처리 성능과 정합성, 격리성 관계
더티 리드, 반복 불가 읽기, 팬텀 리드

직렬화 가능
반복 가능한 읽기 -> 팬텀 리드, mysql InnoDB 기본 격리 수준
커밋된 읽기 -> 팬텀 리드, 반복 불가 읽기
커밋되지 않은 읽기 -> 팬텀 리드, 반복 불가 읽기, 더티 리드
